pkg/standalone-utils/test/EulerWrapping.test.ts

```diff
@@ -0,0 +1,660 @@
+import { expect } from 'chai';
+import { ethers } from 'hardhat';
+import { BigNumber, Contract, ContractReceipt } from 'ethers';
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers';
+
+import Token from '@balancer-labs/v2-helpers/src/models/tokens/Token';
+import TokenList from '@balancer-labs/v2-helpers/src/models/tokens/TokenList';
+import StablePool from '@balancer-labs/v2-helpers/src/models/pools/stable/StablePool';
+
+import { SwapKind, StablePoolEncoder } from '@balancer-labs/balancer-js';
+import * as expectEvent from '@balancer-labs/v2-helpers/src/test/expectEvent';
+import { expectTransferEvent } from '@balancer-labs/v2-helpers/src/test/expectTransfer';
+import { deploy, deployedAt } from '@balancer-labs/v2-helpers/src/contract';
+import { actionId } from '@balancer-labs/v2-helpers/src/models/misc/actions';
+import { MAX_INT256, MAX_UINT256, ZERO_ADDRESS } from '@balancer-labs/v2-helpers/src/constants';
+import { BigNumberish, fp, bn } from '@balancer-labs/v2-helpers/src/numbers';
+import Vault from '@balancer-labs/v2-helpers/src/models/vault/Vault';
+import { Account } from '@balancer-labs/v2-helpers/src/models/types/types';
+import TypesConverter from '@balancer-labs/v2-helpers/src/models/types/TypesConverter';
+import { Dictionary } from 'lodash';
+import {
+  expectChainedReferenceContents,
+  setChainedReferenceContents,
+  toChainedReference,
+} from './helpers/chainedReferences';
+import { sharedBeforeEach } from '@balancer-labs/v2-common/sharedBeforeEach';
+
+describe('EulerWrapping', function () {
+  let DAI: Contract, eDAI: Contract;
+  let mockEulerProtocol: Contract;
+  let senderUser: SignerWithAddress, recipientUser: SignerWithAddress, admin: SignerWithAddress;
+  let vault: Vault;
+  let relayer: Contract, relayerLibrary: Contract;
+
+  before('setup signers', async () => {
+    [, admin, senderUser, recipientUser] = await ethers.getSigners();
+  });
+
+  sharedBeforeEach('deploy Vault', async () => {
+    vault = await Vault.create({ admin });
+
+    mockEulerProtocol = await deploy('MockEulerProtocol');
+
+    DAI = await deploy('v2-solidity-utils/TestToken', { args: ['DAI', 'DAI', 18] });
+
+    const daiAddress = DAI.address;
+    const eulerProtocolAddress = mockEulerProtocol.address;
+
+    eDAI = await deploy('MockEulerToken', {
+      args: ['eDAI', 'eDAI', 18, daiAddress, eulerProtocolAddress],
+    });
+    eDAI.setExchangeRateMultiplier(bn(5e17));
+  });
+
+  sharedBeforeEach('mint tokens to senderUser', async () => {
+    await DAI.mint(senderUser.address, fp(100));
+    await DAI.connect(senderUser).approve(vault.address, fp(100));
+
+    await eDAI.mint(senderUser.address, fp(100));
+    await eDAI.connect(senderUser).approve(eDAI.address, fp(100));
```

Comment:
This approval doesn't make much sense. The user is approving the wrappedToken address to spend wrappedToken. But this should never be necessary because wrappedToken is directly minted and burned via deposit/withdraw.

Instead, let's approve the Vault here so that we don't have to do it later.
```suggestion
    await eDAI.connect(senderUser).approve(vault.address, fp(100));
```